File: 124.binary-tree-maximum-path-sum.py
#
# @lc app=leetcode id=124 lang=python3
#
# [124] Binary Tree Maximum Path Sum
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.max_path_sum = None

    def maxPathSum(self, root: TreeNode) -> int:
        pass
    
    def helper(self,root):
        # if not root:
        #     return 0

        if not root.left and not root.right:
            if self.max_path_sum is None or self.max_path_sum < root.val :
                self.max_path_sum = root.val
            return root.val

        left_subtree_min_value = self.helper(root.left)
        right_subtree_min_value = self.helper(root.right)
        if root.val > self
        


# @lc code=end

File: golang/124.binary-tree-maximum-path-sum.go
/*
 * @lc app=leetcode id=124 lang=golang
 *
 * [124] Binary Tree Maximum Path Sum
 */
package main

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
	max, maxSubtreeSum := helper(root)
	if max < maxSubtreeSum {
		return maxSubtreeSum
	}
	return max
}
func helper(root *TreeNode) (int, int) {
	if root == nil {
		return 0, -9999
	}
	maxLeft, maxPathLeft := helper(root.Left)
	maxRight, maxPathRight := helper(root.Right)

	maxSubtreeSum := maxPathLeft
	if maxSubtreeSum < maxPathRight {
		maxSubtreeSum = maxPathRight
	}

	max := maxPathLeft + maxPathRight + root.Val
	if max < maxLeft {
		max = maxLeft
	}
	if max < maxRight {
		max = maxRight
	}

	return max, maxSubtreeSum + root.Val
}

// @lc code=end
